### PR TITLE
make check-storage-lvm2 nondestructive

### PR DESCRIPTION
**Pull request by martinpitt**
See individual commits for details.

 - [x] builds on top of PR #13650

### CODE REVIEW

**Comment by martinpitt** [comment]
There are some failures/retries on [fedora-32](https://logs-https-cockpit.apps.ci.centos.org/logs/pull-13659-20200305-102952-73b937f5-fedora-32/log.html), [ubuntu-1804](https://logs.cockpit-project.org/logs/pull-13659-20200305-102952-73b937f5-ubuntu-1804/log.html), and [ubuntu-stable](https://logs-https-cockpit.apps.ci.centos.org/logs/pull-13659-20200305-102952-73b937f5-ubuntu-stable/log.html)

**Comment by martinpitt** [comment]
On [debian-stable](https://logs.cockpit-project.org/logs/pull-13659-20200305-180610-3cda1f16-debian-stable/log.html#273) and [ubuntu-1804](https://logs.cockpit-project.org/logs/pull-13659-20200305-180610-3cda1f16-ubuntu-1804/log.html#273), `testSnapshots` still (occasionally) fails with

    Error creating snapshot: Process reported exit code 5:   Volume group "TEST" has insufficient free space (13 extents): 14 required.

File: test/verify/check-storage-lvm2 [hunk 1]
```diff
@@ -33,22 +34,21 @@ class TestStorage(StorageCase):
...
 
-        m.add_disk("50M", serial="DISK1")
-        m.add_disk("50M", serial="DISK2")
-        b.wait_in_text("#drives", "DISK1")
-        b.wait_in_text("#drives", "DISK2")
-
-        dev_1 = "/dev/disk/by-id/scsi-0QEMU_QEMU_HARDDISK_DISK1"
-        dev_2 = "/dev/disk/by-id/scsi-0QEMU_QEMU_HARDDISK_DISK2"
+        dev_1 = self.add_ram_disk()
+        dev_2 = self.add_loopback_disk()
```
Comment (mvollmer) [inline test/verify/check-storage-lvm2]: I think we should use `add_loopback_disk` for both disks, for simplicity.  Or is there a reason that one needs to be a RAM disk, and the other not?

File: test/verify/check-storage-lvm2 [hunk 2]
```diff
@@ -33,22 +34,21 @@ class TestStorage(StorageCase):
...
         m.execute("vgcreate TEST1 %s %s" % (dev_1, dev_2))
+        # just in case the test fails
+        self.addCleanup(m.execute, "vgremove --force TEST 2>/dev/null || true")
         b.wait_in_text("#devices", "TEST1")
         b.wait_in_text("#devices", "/dev/TEST1/")
         b.click('tr:contains("TEST1")')
         b.wait_visible("#storage-detail")
-        b.wait_in_text("#detail-sidebar", "DISK1")
-        b.wait_in_text("#detail-sidebar", "DISK2")
+        b.wait_in_text("#detail-sidebar", "scsi_debug")
```
Comment (mvollmer) [inline test/verify/check-storage-lvm2]: Let's wait for `dev_1` here, and in other places.

File: test/verify/check-storage-lvm2 [hunk 3]
```diff
@@ -21,6 +21,7 @@ import parent
 from storagelib import *
 from testlib import *
 
+@nondestructive
```
Comment (mvollmer) [inline test/verify/check-storage-lvm2]: These tests modify `/etc/fstab`.  They also clean it up, but only when they run all the way to the end.  If a test fails in the middle, we might have something in `/etc/fstab` that will cause the next boot to fail.

Comment (mvollmer) [reply test/verify/check-storage-lvm2]: I guess we should backup/restore `/etc/fstab` and `/etc/crypttab` around all the storage tests.

File: test/verify/storagelib.py [hunk 4]
```diff
@@ -55,6 +55,25 @@ def add_ram_disk(self, size=50):
...
+        simulation of a "real" disk.
+
+        Return the device name.
+        '''
+        dev = self.machine.execute("set -e; F=$(mktemp /var/tmp/loop.XXXX); "
+                                   "dd if=/dev/zero of=$F bs=1M count=%s; "
+                                   "losetup --find --show $F; "
+                                   "rm $F" % size).strip()
+        # right after unmounting the device is often still busy, so retry a few times
+        self.addCleanup(self.machine.execute, "umount {0}; until losetup -d {0}; do sleep 1; done".format(dev), timeout=10)
```
Comment (mvollmer) [inline test/verify/storagelib.py]: Wont this run into the same trouble as the `addCleanup` that used to be in `add_ram_disk`?

File: test/verify/check-storage-lvm2 (hunk 1 above)
Comment (martinpitt) [reply test/verify/check-storage-lvm2]: Yes, there is. loopback devices are a terrible simulation of real disks. They don't appear in the "Drives" list on the overview page, for example. OTOH `scsi_debug` is a pretty faithful emulation of a real hard disk, so I'd much rather test against that as much as possible. With using both we have more diversity in drive types.

File: test/verify/storagelib.py (hunk 4 above)
Comment (martinpitt) [reply test/verify/storagelib.py]: I didn't notice any trouble with that, so as long as it works, it's IMHO more robust to use `addCleanup` at the place where the resource is created. If you prefer, I can move it to `MachineCase.nonDestructiveSetup()` though.

File: test/verify/check-storage-lvm2 (hunk 2 above)
Comment (martinpitt) [reply test/verify/check-storage-lvm2]: I can change the tests to wait for /dev/sda in `#drives`. But here in particular that doesn't work, as `#detail-sidebar` does *not* show the /dev/* names, just the vendor/model name.

Comment (mvollmer) [reply test/verify/check-storage-lvm2]: Ah, it only shows the device name for loop backs?  Good point.

File: test/verify/check-storage-lvm2 (hunk 1 above)
Comment (mvollmer) [reply test/verify/check-storage-lvm2]: Ok, that makes sense, thanks!

File: test/verify/storagelib.py (hunk 4 above)
Comment (mvollmer) [reply test/verify/storagelib.py]: No, let's wait for evidence that this is actually problematic.

File: test/verify/check-storage-lvm2 (hunk 2 above)
Comment (martinpitt) [reply test/verify/check-storage-lvm2]: Done in most places, and I also added waiting for the loop device, and added an extra commit that changes the waiting for all other tests.

File: test/verify/check-storage-lvm2 (hunk 3 above)
Comment (martinpitt) [reply test/verify/check-storage-lvm2]: Good idea, done in a separate commit.

**Comment by martinpitt** [comment]
The new fstab save/restore is broken on [fedora-coreos](https://logs-https-cockpit.apps.ci.centos.org/logs/pull-13659-20200309-120444-9a996835-fedora-coreos/log.html#1-2)

**Pull request merged**